File: app/services/communication_types.py
"""
Service layer for Communication Type management
"""
from app.models import CommunicationType
from app.db import db


def list_communication_types(active_only=True):
    """Get all communication types"""
    query = CommunicationType.query
    if active_only:
        query = query.filter_by(is_active=True)
    return query.order_by(CommunicationType.name).all()


def get_communication_type_by_id(comm_type_id):
    """Get a specific communication type by ID"""
    return CommunicationType.query.get(comm_type_id)


def get_communication_type_by_code(code):
    """Get a specific communication type by code"""
    return CommunicationType.query.filter_by(code=code).first()


def create_communication_type(payload):
    """Create a new communication type"""
    comm_type = CommunicationType(
        code=payload.get("code"),
        name=payload.get("name"),
        description=payload.get("description"),
        requires_device_eui=payload.get("requires_device_eui", False),
        requires_gateway=payload.get("requires_gateway", False),
        supports_remote_control=payload.get("supports_remote_control", False),
        is_active=payload.get("is_active", True),
    )
    db.session.add(comm_type)
    db.session.commit()
    return comm_type


def update_communication_type(comm_type_id, payload):
    """Update an existing communication type"""
    comm_type = get_communication_type_by_id(comm_type_id)
    if not comm_type:
        return None

    if "code" in payload:
        comm_type.code = payload["code"]
    if "name" in payload:
        comm_type.name = payload["name"]
    if "description" in payload:
        comm_type.description = payload["description"]
    if "requires_device_eui" in payload:
        comm_type.requires_device_eui = payload["requires_device_eui"]
    if "requires_gateway" in payload:
        comm_type.requires_gateway = payload["requires_gateway"]
    if "supports_remote_control" in payload:
        comm_type.supports_remote_control = payload["supports_remote_control"]
    if "is_active" in payload:
        comm_type.is_active = payload["is_active"]

    db.session.commit()
    return comm_type


def delete_communication_type(comm_type_id):
    """Soft delete a communication type by setting is_active to False"""
    comm_type = get_communication_type_by_id(comm_type_id)
    if not comm_type:
        return False

    comm_type.is_active = False
    db.session.commit()
    return True

File: app/static/js/settings/communication-types.js
// Communication Type Management JavaScript

function showCreateCommunicationType() {
  document.getElementById("modalTitle").textContent = "Add Communication Type";
  document.getElementById("communicationTypeForm").reset();
  document.getElementById("communicationTypeId").value = "";
  document.getElementById("communicationTypeModal").classList.remove("hidden");
}

function hideCommunicationTypeModal() {
  document.getElementById("communicationTypeModal").classList.add("hidden");
}

async function editCommunicationType(id) {
  try {
    const response = await fetch(`/api/v1/api/communication-types/${id}`);
    if (!response.ok) throw new Error("Failed to fetch communication type");

    const commType = await response.json();

    document.getElementById("modalTitle").textContent = "Edit Communication Type";
    document.getElementById("communicationTypeId").value = commType.id;
    document.getElementById("code").value = commType.code;
    document.getElementById("name").value = commType.name;
    document.getElementById("description").value = commType.description || "";
    document.getElementById("requires_device_eui").checked = commType.requires_device_eui;
    document.getElementById("requires_gateway").checked = commType.requires_gateway;
    document.getElementById("supports_remote_control").checked = commType.supports_remote_control;
    document.getElementById("is_active").checked = commType.is_active;

    document.getElementById("communicationTypeModal").classList.remove("hidden");
  } catch (error) {
    alert("Error loading communication type: " + error.message);
  }
}

async function deactivateCommunicationType(id, name) {
  if (!confirm(`Are you sure you want to deactivate "${name}"?`)) {
    return;
  }

  try {
    const response = await fetch(`/api/v1/api/communication-types/${id}`, {
      method: "DELETE",
    });

    if (!response.ok) throw new Error("Failed to deactivate communication type");

    location.reload();
  } catch (error) {
    alert("Error deactivating communication type: " + error.message);
  }
}

// Form submission
document.getElementById("communicationTypeForm").addEventListener("submit", async (e) => {
  e.preventDefault();

  const formData = new FormData(e.target);
  const id = formData.get("id");

  const payload = {
    code: formData.get("code"),
    name: formData.get("name"),
    description: formData.get("description") || null,
    requires_device_eui: formData.get("requires_device_eui") === "on",
    requires_gateway: formData.get("requires_gateway") === "on",
    supports_remote_control: formData.get("supports_remote_control") === "on",
    is_active: formData.get("is_active") === "on",
  };

  try {
    let response;
    if (id) {
      // Update existing
      response = await fetch(`/api/v1/api/communication-types/${id}`, {
        method: "PUT",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(payload),
      });
    } else {
      // Create new
      response = await fetch("/api/v1/api/communication-types", {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(payload),
      });
    }

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.error || "Failed to save communication type");
    }

    location.reload();
  } catch (error) {
    alert("Error saving communication type: " + error.message);
  }
});

// Close modal on Escape key
document.addEventListener("keydown", (e) => {
  if (e.key === "Escape") {
    hideCommunicationTypeModal();
  }
});
